schema/query/projection_evaluator.go

```diff
@@ -39,6 +39,27 @@ func (p Projection) Eval(ctx context.Context, payload map[string]interface{}, rs
 	return payload, err
 }
 
+func checkStar(p Projection) (bool, error) {
+	hasStar := false
+	s := ""
+	for i := range p {
```

Comment:
This for loop is a bit awkward...

Could not this whole test be just:

```go
func checkStar(p Projection) (bool, error) {
    if len(p) == 1 && p[0].Name == "*" && len p[0].Children == 0 {
        return true, nil
    }
    if len(p) <= 1 {
        return false, nil
    }
    var s, comma string
    for i := range p {
        if p[i].Name == "*"  && len p[i].Children == 0: {
            return false, fmt.Errorf("%s: invalid value: * must be the only field", s)
        }
    }
}
```

This gives less indentation, and a slightly more clear LOS ([Line of Sight](https://medium.com/@matryer/line-of-sight-in-code-186dd7cdea88)).

Reply:
An even smaller version with different errors...
```go
func checkStar(p Projection) bool {
    if len(p) == 1 && p[0].Name == "*" && len p[0].Children == 0 {
        return true
   }
   return false
}
```

Reply:
Otherwise I am now OK with `*` only being allowed as the only parameter, as long as this now expands all references; at least initially. May place nice (end-user-wise) with the suggesrion of using `*` to expand references in schema.Dict and schema.Array

#138

Reply:
Thanks for the suggestions @smyrman. However expanding referenced fields, is what I wanted to avoid in the first place. My indention was to provide extension to this requests:
```
GET http://192.168.0.103:3001/api/clients/59a85126eaba9ebd0180ac55
```
Which will return only non referenced fields, so making this request:
```
GET http://192.168.0.103:3001/api/clients/59a85126eaba9ebd0180ac55?fields=*,visits
```
Is to actually add the `visits` referenced resource, without mentioning all the non referenced fields.

Initially I though this as a low hanging fruit, but it gets more complicated. It is not big deal to enumerate all the fields explicitly, so I think I will just abandon this PR.

Reply:
>  It is not big deal to enumerate all the fields explicitly, so I think I will just abandon this PR.

Alright 👍 Might be wise for now. This is also the way we use it.

Just for my clarification though.

```
GET http://192.168.0.103:3001/api/clients/59a85126eaba9ebd0180ac55?fields=*,visits

```
> Is to actually add the visits referenced resource, without mentioning all the non referenced fields.

This is also what I though was the initial intent. However, my (perhaps wrongful) understanding of the final PR was that it would expand "visits", all other references _and_ maybe (not clear to me) all fields that are hidden by default (Field.Hidden == true) if you did this:

```
GET http://192.168.0.103:3001/api/clients/59a85126eaba9ebd0180ac55?fields=*
```

If we rather want the initially described behaviour, I suppose the following might work:

```go
// expandStar replace a single `*` in projection with all fields at the position of `*`.
// duplicated fields will be handled according to the last-win principal.
expandStart(p *Projection, schema.Fields) {
    ...
}
```
And then no other changes.

Reply:
```
GET http://192.168.0.103:3001/api/clients/59a85126eaba9ebd0180ac55?fields=*
```
The request above would have been equivalent to this:
```
GET http://192.168.0.103:3001/api/clients/59a85126eaba9ebd0180ac55
```
No referenced fields, no hidden fields.
Actually there is no way of getting the hidden fields with the current code. Mentioning them explicitly in the `fields=` parameter results in the following error:
```
GET http://192.168.0.103:3001/api/clients/59a85126eaba9ebd0180ac55?fields=nameTokens
"Invalid `fields` parameter: nameTokens: hidden field"
```

```
Client = schema.Schema{
		Fields: schema.Fields{
			"id": mongo.ObjectIDField,
...
			"nameTokens": {
				Hidden: true,
			},
```

Reply:
Closing the PR now.

Reply:
Thanks for your effort, even if it was not merged this time.